urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/websocket/client/WebSocket.java

```diff
@@ -0,0 +1,423 @@
+package websocket.client;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.net.Socket;
+import java.net.URI;
+
+
+
+/**
+ * An implementation of the websockets interface as
+ * defined in {@link http://dev.w3.org/html5/websockets/}
+ * and specified in {@link http://www.whatwg.org/specs/web-socket-protocol/}.
+ * 
+ */
+public class WebSocket implements Runnable
+{
+	/** The thread that constantly reads from the input stream. */
+	private Thread asyncReadingThread;
+
+	/** The URL the client is connected to. */
+	private final URI url;
+
+	/** The socket. */
+	private Socket socket;
+	
+	/** The input stream. */
+	private InputStream in;
+	
+	/** The output stream. */
+	private PrintStream out;
+
+	/** The ready state attribute. */
+	private ReadyState readyState;
+	
+	/** The open handler. */
+	private OpenHandler openHandler;
+	
+	/** The message handler. */
+	private MessageHandler messageHandler;
+	
+	/** The close handler. */
+	private CloseHandler closeHandler;
+
+	/**
+	 * Gets the url.
+	 *
+	 * @return the url
+	 */
+	public URI getUrl()
+	{
+		return url;
+	}
+
+	/**
+	 * Gets the ready state.
+	 *
+	 * @return the ready state
+	 */
+	public ReadyState getReadyState()
+	{
+		return readyState;
+	}
+
+	/**
+	 * Instantiates a new web socket. 
+	 * Does not automatically connect to the server.
+	 *
+	 * @param url the url
+	 */
+	public WebSocket(final URI url)
+	{
+		this.url = url;
+
+		final String protocol = url.getScheme();
+		final int port = url.getPort();
+
+		if (!protocol.equals("ws"))
+		{
+			throw new IllegalArgumentException("Unsupported protocol: " + protocol);
+		}
+
+		if (port == -1)
+		{
+			throw new IllegalArgumentException("Port undefined!");
+		}
+		
+		readyState = ReadyState.Connecting;
+	}
+
+	/**
+	 * Establishes a websocket connection with the server.
+	 *
+	 */
+	public void connect()
+	{
+		final String host = url.getHost();
+		final int port = url.getPort();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

`getPort()` returns -1 if the port is not set. Check the value returned by `getPort`. Use a default port if the value returned is negative. See the following diffs on GitHub that make this code change: [diff1](https://github.com/apache/skywalking/commit/10d55860dbb4a4c6fec20d54c1a61382bc681481#diff-d4c3e5510bbb6120dc300aa7b2b1c9051d39fd70662af287db112833c7944e11L43), [diff2](https://github.com/opennetworkinglab/onos/commit/6381a24e9d48d4d4e7110cf78b91a6ab47f791d0#diff-7678808b29b9776d35b4ec9d00428db2fa2e2adcfc6ffc7a69812f3ab330fe10L114)

Learn more: [URL.getPort in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#getPort--)